File: 0_팀노트/소수/소인수분해.js
function makePrimes(n) {
  const primes = Array.from({ length: n + 1 }, (v) => true)
  primes.splice(0, 2, false, false)
  for (let num = 2; num < Math.floor(Math.sqrt(n)) + 1; num++) {
    if (primes[num]) {
      for (let i = num * num; i < n + 1; i += num) {
        primes[i] = false
      }
    }
  }
  return primes
}

function makePrimeFactorization(n) {
  const primes = makePrimes(n)
  const primeNumbers = Array.from({ length: n + 1 }, (v, i) => i).filter((x) => primes[x])
  const result = {}
  for (const number of primeNumbers) {
    if (n % number === 0) {
      result[number] = 0
    }
    while (n % number === 0 && n > 1) {
      result[number]++
      n /= number
    }
    if (n === 1) break
  }
  return result
}

File: 0_팀노트/소수/소인수분해.py
from collections import defaultdict
def make_primes(n:int)->list:
    primes = [True] * (n + 1)
    primes[0:2] = [False, False]
    for num in range(2, int(n * 0.5) + 1):
        if primes[num]:
            for i in range(num ** 2, n + 1, num):
                primes[i] = False
    return primes

def make_prime_factorization(n):
    primes = make_primes(n)
    prime_numbers = [x for x in range(2, n + 1) if primes[x]]
    result = defaultdict(int)
    for number in prime_numbers:
        while n % number == 0 and n > 1:
            result[number] += 1
            n //= number
        if n == 1:
            break
    return result


"""
make_prime_factorization(12) 의 출력 값

{
  "2": 2,
  "3": 1,
}
"""